File: platform/lib-java/src/test/java/com/canary/platform/lib/observability/CanaryRestateMeterTest.java
package com.canary.platform.lib.observability;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.Timer;
import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class CanaryRestateMeterTest {

    private SimpleMeterRegistry registry;
    private CanaryMetrics metrics;
    private CanaryRestateMeter meter;

    @BeforeEach
    void setUp() {
        registry = new SimpleMeterRegistry();
        metrics = new CanaryMetrics(registry, "order");
        meter = new CanaryRestateMeter(metrics);
    }

    @Test
    void measureReturnsValueAndRecordsSuccess() throws Exception {
        String result = meter.measure("CheckoutSagaStable.run", () -> "ok");
        assertThat(result).isEqualTo("ok");

        Counter c = registry.find("canary_request_total")
                .tags("substrate", "restate", "outcome", "success", "target", "CheckoutSagaStable.run")
                .counter();
        assertThat(c).isNotNull();
        assertThat(c.count()).isEqualTo(1.0);
    }

    @Test
    void measureRecordsServerErrorWhenBodyThrows() {
        assertThatThrownBy(() -> meter.measure("CheckoutSagaStable.run", () -> {
            throw new RuntimeException("kaboom");
        })).isInstanceOf(RuntimeException.class).hasMessage("kaboom");

        Counter c = registry.find("canary_request_total")
                .tags("substrate", "restate", "outcome", "server_error", "target", "CheckoutSagaStable.run")
                .counter();
        assertThat(c).isNotNull();
        assertThat(c.count()).isEqualTo(1.0);
    }

    @Test
    void measureAlwaysRecordsTimer() throws Exception {
        meter.measure("PaymentVOStable.charge", () -> 1);

        Timer t = registry.find("canary_request_duration_seconds")
                .tags("substrate", "restate", "target", "PaymentVOStable.charge")
                .timer();
        assertThat(t).isNotNull();
        assertThat(t.count()).isEqualTo(1L);
    }
}

File: platform/lib-node/src/__tests__/canary-restate-meter.test.ts
import { Registry } from "prom-client";
import { describe, expect, it } from "vitest";
import { CanaryMetrics } from "../observability/canary-metrics.js";
import { measureRestate } from "../observability/canary-restate-meter.js";

describe("measureRestate", () => {
  it("returns value and records success", async () => {
    const registry = new Registry();
    const metrics = new CanaryMetrics("order", registry);

    const result = await measureRestate(metrics, "CheckoutSagaStable.run", async () => "ok");

    expect(result).toBe("ok");
    const c = await registry.getSingleMetric("canary_request_total")?.get();
    expect(c?.values).toContainEqual(
      expect.objectContaining({
        value: 1,
        labels: expect.objectContaining({
          substrate: "restate",
          outcome: "success",
          target: "CheckoutSagaStable.run",
        }),
      }),
    );
  });

  it("records server_error when body throws and re-throws", async () => {
    const registry = new Registry();
    const metrics = new CanaryMetrics("order", registry);

    await expect(
      measureRestate(metrics, "CheckoutSagaStable.run", async () => {
        throw new Error("boom");
      }),
    ).rejects.toThrow("boom");

    const c = await registry.getSingleMetric("canary_request_total")?.get();
    expect(c?.values).toContainEqual(
      expect.objectContaining({
        value: 1,
        labels: expect.objectContaining({ outcome: "server_error" }),
      }),
    );
  });

  it("always records histogram", async () => {
    const registry = new Registry();
    const metrics = new CanaryMetrics("order", registry);

    await measureRestate(metrics, "PaymentVOStable.charge", async () => 1);

    const h = await registry.getSingleMetric("canary_request_duration_seconds")?.get();
    expect(h?.values.some((v) => v.metricName === "canary_request_duration_seconds_count")).toBe(true);
  });
});
